walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  
+  public final Collection<TimeRange> slots = new ArrayList<>();
+  public final List<TimeRange> bookedTimes = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    if(request.getDuration() >= TimeRange.WHOLE_DAY.duration()){
+        return slots;
+    }else if(events.isEmpty()){
+        slots.add(TimeRange.WHOLE_DAY);
+    }else {
+        for(Event event: events){
+            for(String attendee: request.getAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+
+        if(bookedTimes.size() == 0){
+            slots.add(TimeRange.WHOLE_DAY);
+            return slots;
+        }
+
+        //Sort the booked times by start of day
+        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
+
+        //Merge overlapping or nested timeslots into a single chunk of time
+        for (int i = 0; i < bookedTimes.size() - 1; i++){
+            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
+                if(bookedTimes.get(i).end() > bookedTimes.get(i+1).end()){
```

Comment:
can't you just define endTime = max(bookedTimes.get(i).end(), bookedTimes.get(i+1).end()) and remove the if-else?

Reply:
Yes I can! Resolved